File: mcp-code-intelligence-python/src/mcp_code_intel/orchestration/nested_detection.py
"""Nested orchestrator detection — detect child servers that are orchestrators."""

from __future__ import annotations

import sys

_ORCHESTRATOR_MARKERS = ("find_tools", "execute_dynamic_tool")


def is_nested_orchestrator(tool_names: list[str]) -> bool:
    """Check if a set of tool names indicates a nested orchestrator."""
    return any(marker in tool_names for marker in _ORCHESTRATOR_MARKERS)


def detect_nested(server_name: str, tools: list[dict]) -> bool:
    """Detect and log nested orchestrators from server tool lists."""
    names = [t.get("name", "") for t in tools]
    is_nested = is_nested_orchestrator(names)
    if is_nested:
        print(
            f"[orchestration] Nested orchestrator detected on '{server_name}' — tools accessible via find_tools",
            file=sys.stderr,
            flush=True,
        )
    return is_nested

File: backend/mcp-code-intelligence-nodejs/src/orchestration/nested-detection.ts
/**
 * Nested orchestrator detection — detect child servers that are orchestrators.
 * A server is considered a nested orchestrator if it exposes find_tools or execute_dynamic_tool.
 */

const ORCHESTRATOR_MARKERS = ['find_tools', 'execute_dynamic_tool'];

/** Check if a set of tool names indicates a nested orchestrator. */
export function isNestedOrchestrator(toolNames: string[]): boolean {
  return ORCHESTRATOR_MARKERS.some(marker => toolNames.includes(marker));
}

/** Detect and log nested orchestrators from server tool lists. */
export function detectNested(
  serverName: string,
  tools: Array<{ name: string }>
): boolean {
  const names = tools.map(t => t.name);
  const isNested = isNestedOrchestrator(names);
  if (isNested) {
    console.error(`[orchestration] Nested orchestrator detected on '${serverName}' — tools accessible via find_tools`);
  }
  return isNested;
}
